Modules/Common/TONLikeNFJettonModule.php

```diff
@@ -0,0 +1,265 @@
+
+    final public function pre_process_block($block_id)
+    {
+        if ($block_id === 0) // Block #0 is there, but the node doesn't return data for it
+        {
+            $this->block_time = date('Y-m-d H:i:s', 0);
+            $this->set_return_events([]);
+            $this->set_return_currencies([]);
+            return;
+        }
+
+        $events = [];
+        $currencies_to_process = [];
+        $sort_key = 0;
+
+        $rq_blocks = [];
+        $rq_blocks_data = [];
+        $block_times = [];
+
+        foreach ($this->shards as $shard => $shard_data) 
+        {
+            $this_root_hash = strtoupper($shard_data['roothash']);
+            $this_block_hash = strtoupper($shard_data['filehash']);
+
+            $rq_blocks[] = requester_multi_prepare(
+                $this->select_node(),
+                endpoint: "blockLite?workchain={$this->workchain}&shard={$shard}&seqno={$block_id}&roothash={$this_root_hash}&filehash={$this_block_hash}",
+                timeout: $this->timeout
+            );
+        }
+            
+        $rq_blocks_multi = requester_multi(
+            $rq_blocks,
+            limit: envm($this->module, 'REQUESTER_THREADS'),
+            timeout: $this->timeout,
+            valid_codes: [200]
+        );
+
+        foreach ($rq_blocks_multi as $v)
+            $rq_blocks_data[] = requester_multi_process($v, flags: [RequesterOption::RecheckUTF8]);
+
+        foreach ($rq_blocks_data as $block)
+        {
+            $block_times[] = (int)$block['header']['time'];
+
+            foreach ($block['transactions'] as $transaction)
+            {
+                $transaction['hash'] = strtolower($transaction['hash']);
+
+                if (isset($transaction['messageIn']))
+                {
+                    $messageIn = $transaction['messageIn'][0]; // by default in TON there is only 1 message IN
+
+                    if (isset($messageIn['transfer'])) 
+                    {
+                        if (in_array($transaction['messageIn'][0]['transfer']['transfer_type'], ["transfer", "deploy_nft"])
+                            && $transaction['messageIn'][0]['transfer']['type'] === 'nft')
+                        {
+                            $token_info = $this->api_get_nft_info($transaction['messageIn'][0]['transfer']['token']);
+
+                            $events[] = [
+                                'transaction' => $transaction['hash'],
+                                'currency'    => $token_info['collection_address'],
+                                'address'     => ($transaction['messageIn'][0]['transfer']['from'] === "") ? 'the-void' : $transaction['messageIn'][0]['transfer']['from'],
+                                'sort_key'    => $sort_key++,
+                                'effect'      => '-1',
+                                'extra'       => $token_info['index'],
+                                'extra_indexed' => $transaction['messageIn'][0]['transfer']['token'],
```

Comment:
Could you please comment on what's contained in `extra` and `extra_indexed`?

If `extra` contains unique NFT id, the extra model should be set like this: `public ?ExtraDataModel $extra_data_model = ExtraDataModel::Identifier;`

Reply:
In TON every NFT has own address, but it also has an own index. Unfortunately, contracts realizations do not always support getting NFT address by index or index can be something else, so for data consistency it's better to collect NFT address as well. However, the comment about `$extra_data_model` is right, will update this